### PR TITLE
[2424] Ajax trainee filtering

### PR DESCRIPTION
**Pull request by AbigailMcP**
### Context

https://trello.com/c/BPsUCM8i/2424-l-ajax-trainee-list-filtering

### Changes proposed in this pull request

Adds Ajax filtering to our existing filters on the trainee index page.

**Things to note:**

1. I've removed the `PaginatedFilter` and `Filter` component.
  - All the `PaginatedFilter` component did was pass variables through to the `Filter` component and include a `.scss` file
  - All the `Filter` component did was encapsulate some helpers and render html slots that were defined in the top `index.html.erb`.
All this felt needlessly complicated and the nesting made the Ajax implementation much harder (since there were no defined partials to render dynamically)

2. I've moved the couple of methods from `Filters::View` into a `FiltersHelper` module, and included the html in partials.

3. These partials are rendered via Ajax and targeted via their `js-` prefixed IDs.

### TODO
- ~How to update the page title (help!)~
- ~Think about accessibility (help!)~
- ~Jest tests - I will add!~

### Guidance to review
- Filter trainees on `/trainees`
- Check that:
- [ ] The correct trainees are returned
- [ ] The trainee count is correct
- [ ] The correct selected filters are displayed
- [ ] The url is updated
- [ ] The page title is updated
- [ ] The submit button is removed
- [ ] The trainee export is correct
- [ ] The pagination is correct, especially the counts
- [ ] The sorting retains the filters
- [ ] When you disable JS, everything works the way it did before
- [ ] Using VoiceOver (or equivalent), a user is notified that something has changed when the filters are updated
- [ ] The browser back button works 😬

### CODE REVIEW

File: app/helpers/filter_helper.rb [hunk 1]
```diff
@@ -0,0 +1,38 @@
+# frozen_string_literal: true
```
Comment (AbigailMcP) [inline app/helpers/filter_helper.rb:1]: Copied directly from our `Filter` component, no new code.

File: app/views/trainees/_filters.html.erb [hunk 2]
```diff
@@ -0,0 +1,89 @@
+<form method="get" id="js-live-filter">
```
Comment (AbigailMcP) [inline app/views/trainees/_filters.html.erb:1]: Pulled from `index.rb` - no new code except for addition of `js` ids to use as selectors.

File: app/views/trainees/_results.html.erb [hunk 3]
```diff
@@ -0,0 +1,57 @@
+<div class="app-records-actions">
```
Comment (AbigailMcP) [inline app/views/trainees/_results.html.erb:1]: Pulled from index.rb - no new code.

**Comment by AbigailMcP** [comment]
[[L] Ajax trainee list filtering](https://trello.com/c/BPsUCM8i/2424-l-ajax-trainee-list-filtering)

File: app/webpacker/scripts/live_search.js [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+    this.selectedFiltersDiv.innerHTML = response.selected_filters
+    this.traineeCount.innerHTML = response.trainee_count
+  }
+
+  updatePage () {
+    return $.ajax({
+      url: '/trainees',
+      dataType: 'json',
+      data: this.state
+    }).done(response => { this.displayResults(response) })
```
Comment (defong) [inline app/webpacker/scripts/live_search.js:38]:
```suggestion
    }).done(this.displayResults)
```

File: app/webpacker/scripts/live_search.js [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+  saveState () {
+    const serializedForm = $(this.form).serializeArray()
+    this.state = serializedForm.filter(field => field.value !== '')
+  }
+
+  formChange () {
+    this.saveState()
+    const pageUpdated = this.updatePage()
+    pageUpdated.done(() => { this.updateUrl() })
```
Comment (defong) [inline app/webpacker/scripts/live_search.js:49]:
```suggestion
    this.updatePage().done(this.updateUrl)
```

File: app/webpacker/scripts/live_search.js (hunk 4 above)
Comment (AbigailMcP) [reply app/webpacker/scripts/live_search.js:38]: i do this every. single. time. will i ever learn?

Comment (AbigailMcP) [reply app/webpacker/scripts/live_search.js:38]: needs a `bind` though 😞

File: app/views/trainees/index.html.erb [hunk 6]
```diff
@@ -14,140 +19,40 @@
...
-    <% end %>
 
-    <% if @completed_trainees.any? %>
-      <div class="govuk-!-margin-bottom-8 app-non-draft-records">
-        <div class="govuk-grid-row">
-          <div class="govuk-grid-column-one-half">
-            <h2 class="govuk-heading-m">Records</h2>
+        <!-- Div made focusable to 'catch' focus from filters. see https://github.com/DFE-Digital/apply-for-teacher-training/pull/2640 -->
+        <div class="moj-filter__content" tabindex="-1">
+          <div id='js-selected-filters' aria-live="polite">
```
Comment (AbigailMcP) [inline app/views/trainees/index.html.erb:34]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/ARIA_Live_Regions

**Comment by AbigailMcP** [comment]
I've added an `aria-live=polite` to the selected filters div to announce to the user that the filters have changed, but I'm not sure what's best. Would it be nicer to announce the count? cc @edwardhorsford 

This is the selected filters announcement:

Screenshot 2021-08-11 at 17 39 22

**Comment by ralph-hawkins** [comment]
> I've added an `aria-live=polite` to the selected filters div to announce to the user that the filters have changed, but I'm not sure what's best. Would it be nicer to announce the count? cc @edwardhorsford

I think its clearer if voice over announces the number of results here as way to give the user an indication of the update to the page

**Comment by ralph-hawkins** [comment]
It looks like this introduces a bug where the link to close filters at mobile widths is missing, so a user can't escape the filter menu after they've opened it.

**Comment by AbigailMcP** [comment]
> > I've added an `aria-live=polite` to the selected filters div to announce to the user that the filters have changed, but I'm not sure what's best. Would it be nicer to announce the count? cc @edwardhorsford
> 
> I think its clearer if voice over announces the number of results here as way to give the user an indication of the update to the page

I've changed the aria-live so that it announces the page heading 'Trainee records (4)'.

**Comment by AbigailMcP** [comment]
> It looks like this introduces a bug where the link to close filters at mobile widths is hidden, so a user can't escape the filter menu after they've opened it.

Thank you! Fixed now.

File: app/webpacker/scripts/live_filter.spec.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
+import $ from 'jquery'
+import LiveFilter from './live_filter'
+
+describe('LiveFilter', () => {
+  beforeEach(() => {
+    const $form = $(`<form method="get" id="js-live-filter">
+    <input type="submit" name="commit" value="Apply filters" id="js-submit">
```
Comment (darokel) [inline app/webpacker/scripts/live_filter.spec.js:7]:
```suggestion
      <input type="submit" name="commit" value="Apply filters" id="js-submit">
      <input type="checkbox" name="level[]" id="level-early_years" value="early_years">
      <input type="checkbox" name="level[]" id="level-primary" value="primary">
      <select name="subject" id="subject" class="govuk-select"><option value="All subjects">All subjects</option>
        <option value="ancient hebrew">Ancient hebrew</option>
        <option value="applied biology">Applied biology</option>
        <option value="applied chemistry">Applied chemistry</option>
      </select>
```

File: app/webpacker/scripts/live_filter.spec.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
+import $ from 'jquery'
+import LiveFilter from './live_filter'
+
+describe('LiveFilter', () => {
```
Comment (darokel) [inline app/webpacker/scripts/live_filter.spec.js:4]: ❤️

**Review by darokel — APPROVED** [review]

**Comment by jhenderson** [comment]
This is a cool feature. Nicely done @AbigailMcP

File: app/views/trainees/_action_bar.html.erb [hunk 9]
```diff
@@ -0,0 +1,12 @@
...
+    <div class="app-export--link">
+      <%= govuk_link_to(
+        I18n.t("views.trainees.index.export"),
+        trainees_path(filter_params.merge(format: "csv")),
+        class: "app-trainee-export",
+      ) %>
+    </div>
+  <% end %>
+  <%= render SortLinks::View.new %>
+<% end %>
```
Comment (defong) [inline app/views/trainees/_action_bar.html.erb:12]: line ending

**Review by defong — APPROVED** [review]

File: app/views/trainees/_action_bar.html.erb (hunk 9 above)
Comment (AbigailMcP) [reply app/views/trainees/_action_bar.html.erb:12]: okkkkk

**Review by ralph-hawkins — APPROVED** [review]
nice 👍🏻

**Pull request merged**